### DISCUSSION TITLE
Typegoose and Type-GraphQL name confliction

### DISCUSSION THREAD

**Discussion by arashi-dev** [MichalLytek/type-graphql#1128] [Q&A]
Hello.

I have a problem with `FieldResolver`. It looks like type-graphql doesn't call `FieldResolver` whenever the parent query (like `findProduct`) returns an object which have a property with the same name of `FieldResolver`.
Sorry for my poor english.

This is my simplified code:
```ts
@Resolver(() => Product)
export class ProductResolver {
    @Query(() => Product, { nullable: true })
    async findProduct(): Promise<(Product | null)> {
        return await ProductModel.findById("some id")
    }

    @FieldResolver(() => [Color])
    async colors(@Root() root: Product): Promise<Color[]> {
        console.log("colors property");
        
        // query colors
        const results = await ColorModel.find({ _id: { $in: root.colors as ObjectId[] } })
        console.log(results)
        return results
    }
}
```
And the simplified product ObjectType:

```ts
@ObjectType()
export class Product {
    @Field(() => ObjectId)
    readonly _id: ObjectId;

    @Field(() => [Color], { defaultValue: [] })
    @Prop({ ref: () => "Color", default: [] }, WhatIsIt.ARRAY)
    colors?: Ref<Color>[];
}
```

and this is the error I get:
```ts
"Cannot return null for non-nullable field Color.hex."
```
<hr/>

So I changed the type-graphql colors property name to `colors2` while the typegoose colors property name was still `colors` and it was fixed 🎉 
Looks like type-graphql ignores `colors FieldResolver` because `findProduct` already has a `colors` property.

What do you suggest? should I have separated names or there is a cleaner way?

**Comment by MichalLytek** [comment]
You need to return `.toObject()` values, otherwise TypeGraphQL internals fails to convert typegoose document into `@Root() root: Product`

**Reply by arashi-dev** [reply]
Thanks for your response.
I tried but it didn't work

```ts
return (await ProductModel.findById(productId)).toObject()
```

everything is ok for products that have empty arrays of `colors` but the products that have at least 1 `ObjectId` in `colors` array, get errors.

**Reply by arashi-dev** [reply]
I found out where the problem comes from.

in ObjectType I told that `colors` prop should return an array of `Color ObjectType`. so, when I return ObjectId, it gives me error.
Now, the solution is to create a union type which allows `Color` and `ObjectId` type or there is an easier way? 
because I don't think this is the a good solution

**Reply by arashi-dev** [reply]
and still have that problem too. `findProduct` ignores `colors FieldResolver` because it already have a property named 'colors'.
This is my test. I returned `colors` with an empty array.
```ts
result.colors = []
return result
```
and it didn't call `colors FieldResolver`.

so there is 2 problems.
1. `findProduct` ignores `@FieldResolver` when returning a property that has the same name of `@FieldResolver`.
2. `findProduct` throws an error when returning an array of ObjectId because I told `@Field` that is should return `Color @ObjectType`. It is fine but I think it should first call `@FieldResolver` method and then, validate the returned data.

**Reply by arashi-dev** [reply]
sorry for asking but are you working on it, sir? or I should open an issue for this? because I still have this problem.
